### PR TITLE
ext-authz: fix missing UAEX flag on Denied CheckResponse

### LINKED ISSUE #18964: Bug: Missing Access Log UAEX flag for ext-authz denied requests
*Title*: Missing Access Log UAEX flag for ext-authz denied requests

*Description*:

When upgrading from Envoy 1.14.X to newer versions, the `UAEX` Access Log flag is no longer present.

We would expect logs to show:
```
[2021-11-10T17:49:25.117Z] "GET / HTTP/1.1" 401 UAEX 0 46 2 - "-" "curl/7.78.0" "b8837414-03e2-44c2-985a-9ea679f57183" "example.com:8443" "-"
```
but instead we see `-`:
```
[2021-11-10 09:49:25.119][4652345][trace][ext_authz] [external/envoy/source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:52] Received CheckResponse: status {
  code: 16
}
denied_response {
  status {
    code: Unauthorized
  }
  body: "Authorization Header malformed or not provided"
}

[2021-11-10 09:49:25.119][4652345][trace][filter] [external/envoy/source/extensions/filters/http/ext_authz/ext_authz.cc:237] [C6][S1434089409734892564] ext_authz filter rejected the request. Response status code: '401
[2021-11-10 09:49:25.119][4652345][trace][filter] [external/envoy/source/extensions/filters/http/ext_authz/ext_authz.cc:261] [C6][S1434089409734892564] ext_authz filter added header(s) to the local response:
[2021-11-10T17:49:25.117Z] "GET / HTTP/1.1" 401 - 0 46 2 - "-" "curl/7.78.0" "b8837414-03e2-44c2-985a-9ea679f57183" "example.com:8443" "-"
```

*Repro*:

The bug was introduced in https://github.com/envoyproxy/envoy/commit/b0c3683a9e9bc1ce15a58d385353f2525b3ff515

*Config*:
```yaml
admin:
  access_log_path: /dev/null
  address:
    socket_address:
      address: 127.0.0.1
      port_value: 9001
node:
  cluster: foo
  id: test-id

static_resources:
  listeners:
  - listener_filters:
    - name: "envoy.filters.listener.tls_inspector"
      typed_config: {}
    address:
      socket_address:
        address: 0.0.0.0
        port_value: 8443
    filter_chains:
    - filter_chain_match:
        server_names: ["example.com", "www.example.com"]
      transport_socket:
        name: envoy.transport_sockets.tls
        typed_config:
          "@type": type.googleapis.com/envoy.api.v2.auth.DownstreamTlsContext
          common_tls_context:
            validation_context:
              trusted_ca:
                filename: "/etc/envoy/certs/out/myCA.crt"
            tls_certificates:
            - certificate_chain: { filename: "/etc/envoy/certs/out/example.com.crt" }
              private_key: { filename: "/etc/envoy/certs/out/example.com.key" }
      filters:
      - name: envoy.http_connection_manager
        typed_config:  
          "@type": type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager
          stat_prefix: ingress_http
          codec_type: AUTO
          access_log: # configure logging
            name: envoy.file_access_log
            config:
              path: /dev/stdout
          route_config:
            name: local_route
            virtual_hosts:
            - name: local_service
              domains: ["*"]
              routes:
              - match:
                  prefix: "/"
                route:
                  cluster: service_backend
                per_filter_config:
                  envoy.ext_authz:
                    check_settings:
                      context_extensions:
                        foo: bar
          http_filters: 
          - name: envoy.ext_authz
            config:
              failure_mode_allow: false
              grpc_service:
                envoy_grpc:
                  cluster_name: ext-authz
                timeout: 0.5s
          - name: envoy.router

  clusters:
  - name: service_backend
    connect_timeout: 2s
    type: STRICT_DNS
    dns_lookup_family: V4_ONLY
    lb_policy: ROUND_ROBIN
    hosts:
    - socket_address:
        address: host.docker.internal
        port_value: 11000
  - name: ext-authz
    type: STRICT_DNS
    http2_protocol_options: {}
    load_assignment:
      cluster_name: ext-authz
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: host.docker.internal
                port_value: 50050
    connect_timeout: 0.25s
```

*Logs*:

```
[2021-11-10 09:49:25.117][4652345][trace][filter] [external/envoy/source/extensions/filters/http/ext_authz/ext_authz.cc:69] [C6][S1434089409734892564] ext_authz filter calling authorization server
[2021-11-10 09:49:25.118][4652345][trace][ext_authz] [external/envoy/source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:45] Sending CheckRequest: attributes {
  source {
    address {
      socket_address {
        address: "127.0.0.1"
        port_value: 65230
      }
    }
  }
  destination {
    address {
      socket_address {
        address: "127.0.0.1"
        port_value: 8443
      }
    }
    principal: "www.example.com"
  }
  request {
    time {
      seconds: 1636566565
      nanos: 117749000
    }
    http {
      id: "1434089409734892564"
      method: "GET"
      headers {
        key: ":authority"
        value: "example.com:8443"
      }
      headers {
        key: ":method"
        value: "GET"
      }
      headers {
        key: ":path"
        value: "/"
      }
      headers {
        key: "accept"
        value: "*/*"
      }
      headers {
        key: "user-agent"
        value: "curl/7.78.0"
      }
      headers {
        key: "x-forwarded-proto"
        value: "https"
      }
      headers {
        key: "x-request-id"
        value: "b8837414-03e2-44c2-985a-9ea679f57183"
      }
      path: "/"
      host: "example.com:8443"
      protocol: "HTTP/1.1"
    }
  }
  context_extensions {
    key: "foo"
    value: "bar"
  }
  metadata_context {
  }
}

[2021-11-10 09:49:25.119][4652345][trace][ext_authz] [external/envoy/source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:52] Received CheckResponse: status {
  code: 16
}
denied_response {
  status {
    code: Unauthorized
  }
  body: "Authorization Header malformed or not provided"
}

[2021-11-10 09:49:25.119][4652345][trace][filter] [external/envoy/source/extensions/filters/http/ext_authz/ext_authz.cc:237] [C6][S1434089409734892564] ext_authz filter rejected the request. Response status code: '401
[2021-11-10 09:49:25.119][4652345][trace][filter] [external/envoy/source/extensions/filters/http/ext_authz/ext_authz.cc:261] [C6][S1434089409734892564] ext_authz filter added header(s) to the local response:
[2021-11-10T17:49:25.117Z] "GET / HTTP/1.1" 401 - 0 46 2 - "-" "curl/7.78.0" "b8837414-03e2-44c2-985a-9ea679f57183" "example.com:8443" "-"
```

### PR DESCRIPTION
**Pull request by pims**
This fixes a bug in which the UAEX flag is not set prior to calling
`callback->sendLocalReply(...)`.

Commit Message: ext-authz: fix missing UAEX flag on Denied CheckResponse
Additional Description: See #18964
Risk Level: low
Testing: manual
Docs Changes: n/a
Release Notes: ext-authz: fix missing UAEX flag on Denied CheckResponse

Fixes #18964

### CODE REVIEW

**Comment by yanavlasov** [comment]
@pims thanks for fixing this. Can you also add a test case for the fixed problem, please?

**Comment by pims** [comment]
@yanavlasov thanks for taking a look at this PR.

I've updated the tests to reflect the new call sequence.
To expand a little bit on why this PR exists.

Original code

```cpp
decoder_callbacks_->sendLocalReply(response->status_code, …);
decoder_callbacks_->streamInfo().setResponseFlag(
        StreamInfo::ResponseFlag::UnauthorizedExternalService);
```
My understanding is that the code intends to set the `UnauthorizedExternalService` response flag, but that flag is never visible in Access Logs.

After re-ordering the calls, the unit tests were failing.
In order to understand why, I looked at the fault filter for inspiration/confirmation that the expected behavior is:

```cpp
// source/extensions/filters/http/fault/fault_filter.cc

decoder_callbacks_->streamInfo().setResponseFlag(StreamInfo::ResponseFlag::FaultInjected);
decoder_callbacks_->sendLocalReply(http_status_code, "fault filter abort", …);

// test/extensions/filters/http/fault/fault_filter_test.cc
EXPECT_CALL(decoder_filter_callbacks_.stream_info_,
            setResponseFlag(StreamInfo::ResponseFlag::DelayInjected));
EXPECT_EQ(Http::FilterHeadersStatus::StopIteration,
            filter_->decodeHeaders(request_headers_, false));
```

Following the above pattern, tests are now passing.

```cpp
EXPECT_CALL(filter_callbacks_.stream_info_,
            setResponseFlag(Envoy::StreamInfo::ResponseFlag::UnauthorizedExternalService));
EXPECT_EQ(Http::FilterHeadersStatus::StopAllIterationAndWatermark,
            filter_->decodeHeaders(request_headers_, false));
```

I'm very new to the Envoy code base, so please let me know if my assumptions aren't correct.
Thanks again for reviewing this PR.

**Comment by pims** [comment]
I forced push to fix the DCO as instructed in https://github.com/envoyproxy/envoy/blob/main/CONTRIBUTING.md#fixing-dco

**Comment by pims** [comment]
@dio @soulxu let me know if there’s anything I can do to make this PR worthy of merging.

File: source/extensions/filters/http/ext_authz/ext_authz.cc [hunk 1]
```diff
@@ -346,6 +346,8 @@ void Filter::onComplete(Filters::Common::ExtAuthz::ResponsePtr&& response) {
       config_->httpContext().codeStats().chargeResponseStat(info, false);
     }
 
+    decoder_callbacks_->streamInfo().setResponseFlag(
```
Comment (dio) [inline source/extensions/filters/http/ext_authz/ext_authz.cc:349]: This makes sense. 

Optional: Mind to add a comment for future readers, that is why we set it prior to calling `sendLocalReply`.

**Review by dio — APPROVED** [review]
Looks good. Thank you. A small optional request:

File: source/extensions/filters/http/ext_authz/ext_authz.cc (hunk 1 above)
Comment (soulxu) [reply source/extensions/filters/http/ext_authz/ext_authz.cc:349]: also can put setResponseFlag and sendLocalReply into a inline method, but currently this is enough for me.

**Review by soulxu — APPROVED** [review]
This is LGTM

**Comment by pims** [comment]
@dio added a note to the release notes about this bug fix, following the format you used for the network filter bug fix.

@soulxu thanks for the review.

**Review by soulxu — APPROVED** [review]
still LGTM

**Comment by pims** [comment]
@dio gentle nudge. Anything else you think should be added to this PR for it to be merged?

I have found a similar bug in another filter, and was hoping to follow this PR’s template when submitting a bug fix.

**Review by dio — APPROVED** [review]
Looks good. Thanks!

**Pull request merged**